libp2p/network/url_conversion_test.go

```diff
@@ -173,33 +182,136 @@ func TestServiceURLFromLibp2pMultiaddr_Error(t *testing.T) {
-	// 		"multi.AAAA.example",
-	// 	},
-	// }
+func TestGetPeerIP_SingleRecord_Success(t *testing.T) {
+	testCases := []struct {
+		name       string
+		hostname   string
+		recordType dnsRecordType
+		expectedIP string
+	}{
+		{
+			"A record",
+			"single.a.example",
+			aRecord,
+			"10.0.0.1",
+		},
+		{
+			"AAAA record",
+			"single.aaaa.example",
+			quadARecord,
+			"fc00::1",
+		},
+	}
+
+	// Setup mock DNS
+	zones := make(map[string]mockdns.Zone)
+	for _, testCase := range testCases {
+		// Fully qualified domain name
+		// (see: https://en.wikipedia.org/wiki/Fully_qualified_domain_name)
+		fqdn := fmt.Sprintf("%s.", testCase.hostname)
+
+		switch testCase.recordType {
+		case aRecord:
+			zones[fqdn] = mockdns.Zone{
+				A: []string{testCase.expectedIP},
+			}
+		case quadARecord:
+			zones[fqdn] = mockdns.Zone{
+				AAAA: []string{testCase.expectedIP},
+			}
+		}
+	}
+	closeDNSMock := prepareDNSResolverMock(t, zones)
+
+	// Run tests
+	for _, testCase := range testCases {
+		t.Run(testCase.name, func(t *testing.T) {
+
+			actualIP, err := getPeerIP(testCase.hostname)
+			require.NoError(t, err)
+			require.Equal(t, testCase.expectedIP, actualIP.String())
+		})
+
+	}
+	closeDNSMock()
+}
+
+func TestGetPeerIP_MultipleRecord_Success(t *testing.T) {
+	testCases := []struct {
+		name        string
+		hostname    string
+		recordType  dnsRecordType
+		expectedIPs []string
+	}{
+		{
+			"A records",
+			"multi.a.example",
+			aRecord,
+			[]string{
+				"10.0.0.2",
+				"10.0.0.3",
+				"10.0.0.4",
+				"10.0.0.5",
+			},
+		},
+		{
+			"AAAA records",
+			"multi.aaaa.example",
+			quadARecord,
+			[]string{
+				"fc00::1",
+				"fc00::2",
+				"fc00::3",
+				"fc00::4",
+			},
+		},
+	}
+
+	// Setup mock DNS
+	zones := make(map[string]mockdns.Zone)
+	for _, testCase := range testCases {
+		// Fully qualified domain name
+		// (see: https://en.wikipedia.org/wiki/Fully_qualified_domain_name)
+		fqdn := fmt.Sprintf("%s.", testCase.hostname)
+
+		switch testCase.recordType {
+		case aRecord:
+			zones[fqdn] = mockdns.Zone{
+				A: testCase.expectedIPs,
+			}
+		case quadARecord:
+			zones[fqdn] = mockdns.Zone{
+				AAAA: testCase.expectedIPs,
+			}
+		}
+	}
+	closeDNSMock := prepareDNSResolverMock(t, zones)
+
+	// Run tests
+	for _, testCase := range testCases {
+		t.Run(testCase.name, func(t *testing.T) {
+			seenIPCounts := make(map[string]uint)
+			maxAttempts := len(testCase.expectedIPs) * 100 // arbitrary scalar
+			for i := 0; i < maxAttempts; i++ {
+				// Break if all IPs already seen
+				if len(seenIPCounts) == len(testCase.expectedIPs) {
+					break
+				}
+
+				actualIP, err := getPeerIP(testCase.hostname)
+				require.NoError(t, err)
+
+				seenIPCounts[actualIP.String()]++
+			}
+			var seenIPs []string
+			for ip := range seenIPCounts {
+				seenIPs = append(seenIPs, ip)
+			}
+			require.ElementsMatchf(t, seenIPs, testCase.expectedIPs, "expected and seen IPs don't match")
+		})
+
+	}
+	closeDNSMock()
```

Comment:
Why not defer after 
```
closeDNSMock := prepareDNSResolverMock(t, zones)
```
?

Reply:
No reason, nice catch - thanks! :raised_hands: